src/cljam/io/csi.clj

```diff
@@ -0,0 +1,72 @@
+(ns cljam.io.csi
+  "Reader of a CSI format file."
+  (:require [cljam.io.util.bgzf :as bgzf]
+            [cljam.io.util.lsb :as lsb]
+            [cljam.io.util.bin :as util-bin])
+  (:import java.util.Arrays
+           [java.io DataInputStream IOException]
+           [cljam.io.util.chunk Chunk]))
+
+(deftype CSI [n-ref min-shift depth bidx loffset]
+  util-bin/IBinningIndex
+  (get-chunks [_ ref-idx bins]
+    (vec (mapcat (get bidx ref-idx) bins)))
+  (get-min-offset [_ ref-idx beg]
+    (if-let [min-offsets (rsubseq (get loffset ref-idx) <= beg)]
+      (second (first min-offsets))
+      0))
+  (get-depth [_]
+    depth)
+  (get-min-shift [_]
+    min-shift))
+
+(def ^:const ^:private csi-magic "CSI\1")
+
+(defn- read-chunks!
+  [rdr]
+  (->> #(Chunk. (lsb/read-long rdr) (lsb/read-long rdr))
+       (repeatedly (lsb/read-int rdr))
+       vec))
+
+(defn- read-bin-index
+  [rdr]
+  (->> #(hash-map
+         :bin (lsb/read-int rdr)
+         :loffset (lsb/read-long rdr)
+         :chunks (read-chunks! rdr))
```

Comment:
In current implementation of Clojure, function arguments are certainly evaluated from left to right, but it's not very well known style.
I think using let form before `hash-map` might be more reasonable for side-effects. Please correct me if I'm wrong.
https://clojure.org/reference/evaluation